staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/rule.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filterconfig
+
+import (
+	"strings"
+
+	fctypesv1a1 "k8s.io/api/flowcontrol/v1alpha1"
+	"k8s.io/apiserver/pkg/authentication/serviceaccount"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/endpoints/request"
+)
+
+// Tests whether a given request and FlowSchema match.  Nobody mutates
+// either input.
+func matchesFlowSchema(digest RequestDigest, flowSchema *fctypesv1a1.FlowSchema) bool {
+	for _, policyRule := range flowSchema.Spec.Rules {
+		if matchesPolicyRule(digest, &policyRule) {
+			return true
+		}
+	}
+	return false
+}
+
+func matchesPolicyRule(digest RequestDigest, policyRule *fctypesv1a1.PolicyRulesWithSubjects) bool {
+	if !matchesASubject(digest.User, policyRule.Subjects) {
+		return false
+	}
+	if digest.RequestInfo.IsResourceRequest {
+		return matchesAResourceRule(digest.RequestInfo, policyRule.ResourceRules)
+	}
+	return matchesANonResourceRule(digest.RequestInfo, policyRule.NonResourceRules)
+}
+
+func matchesASubject(user user.Info, subjects []fctypesv1a1.Subject) bool {
+	for _, subject := range subjects {
+		if matchesSubject(user, subject) {
+			return true
+		}
+	}
+	return false
+}
+
+func matchesSubject(user user.Info, subject fctypesv1a1.Subject) bool {
+	switch subject.Kind {
+	case fctypesv1a1.SubjectKindUser:
+		return subject.User != nil && (subject.User.Name == fctypesv1a1.NameAll || subject.User.Name == user.GetName())
+	case fctypesv1a1.SubjectKindGroup:
+		return subject.Group != nil && (containsString(subject.Group.Name, user.GetGroups(), fctypesv1a1.NameAll))
```

Comment:
I think this is still wrong: it will match a request from a user in a group named `*` even if the subject group is not a wildcard, but will not match a user not in a group named `*` if the subject group is a wildcard.

Reply:
Oh, right, this call is opposite all the others.  Thanks, will fix.

Reply:
Need better testing too.

Reply:
There is another odd thing in this vicinity.  Some new options slipped in during the evolution.  Originally the only way to match "any user" was to match on a list of groups that includes both `system:authenticated` and `system:unauthenticated`.  You can see this painful cirumlocution in exhibited configurations.  Now we have two simpler choices: match on username `*` or user group name `*`.  Frankly, I am for these.  Their exclusion seems odd.

Reply:
Fixed, after making sure testing catches this bug.